common/PriorityCache: First Step toward priority based caching

markhpc opened the pull request:
This PR depends on either the upstream rocksdb PR here:

https://github.com/facebook/rocksdb/pull/3732

or a PR against our fork here:

https://github.com/ceph/rocksdb/pull/39

# Problem:

The way that we currently cache data in bluestore is suboptimal.  Bluestore relies heavily on 3 different kinds of cache:

1) metadata cache
2) data cache
3) rocksdb block cache

Currently, we implement a scheme where the user specifies a total amount of cache in bytes that bluestore should allocate and specifies ratios that should be allocated to each cache.  Uunfortunately there are corner cases where this scheme doesn't work well.  One example is during heavy RGW workloads where the total DB size becomes very large, and it's important to give rocksdb enough memory to cache all of the indexes and bloom filters.  Users are required to change the ratios assigned to each cache as the number of objects in RGW grows.  To help alleviate this, we implemented a "minimum" KV cache setting so that in most cases the users could just leave the default ratios while still making sure a minimum amount of memory was assigned to the KV store, however this meant that users were confused when different ratios were set than they specified.

Another issue is that during RBD workloads, you may need more or less memory than is assigned to the metadata cache to fully cache all bluestore onode metadata, and the amount of cache needed changes based on how much RBD data is present.  In master, bluestore can change the amount of memory allocated to data and metadata depending on how much is used.  This helps, but does not necessarily avoid the scenario where there is not enough cache to avoid caching all onodes and onodes end up double cached in rocksdb's cache and bluestore's cache.  It also further confuses the users, especially since even under memory pressure bluestore's caches tend to over-allocate memory to the metadata cache beyond what the user specified.

![master_behavior](https://user-images.githubusercontent.com/1286295/40058491-e37ef3f6-5816-11e8-8e24-813edfe97915.png)

Ultimately, for optimal behavior, the ratios of caches must change to match the current workload, and so it's impossible for both us as developers and for users to specify static ratios that work well under all scenarios.  The result of all of this is that users have a hard time understanding how these different options and behaviors interact, how their settings ultimately affect the amount of memory assigned to different caches, and how they should configure these settings on their cluster.

# What this PR Does

This PR introduces the concept of cache autotuning in Ceph, along with laying the foundation for a concept called priority based caching.  

# Autotuning Disabled

When autotuning is disabled, we should make bluestore's behavior as transparent to the user as possible.  IE given a certain cache size and certain ratios, set the cache sizes accordingly and don't mess with them beyond that.  This may result in less (or more) optimal behavior than current master, but does exactly what the user requests (and leaves it up to them to find the right ratios).

# Autotuning Enabled

When autotuning is enabled, bluestore will periodically query each cache for how much memory it would like for a given priority (typically how much memory is currently being used plus some extra space for expansion).  It will either allocate this amount or the user-defined fair ratio of remaining memory depending on which is larger.  If any caches don't get as much memory for this priority as they want and there is still memory available, that cache may get additional memory assigned to it to fulfill it's current priority needs before moving on to lower priorities.  Currently the only high priority items are rocksdb's index and filter blocks in rocksdb's high priority cache pool.  This PR lays the ground work for supporting an arbitrary number of generic priorities however.  A future PR for example may implement the concept of binned LRUs where each cache segments their LRU into different priority bins based on age.  If for example, the KV cache generally has older items than bluestore's onode cache, it would get a lower proportion of the available memory.

# RBD Behavior - Autotuning Disabled

![wip-bs-cache-autotune-40-40-20-false](https://user-images.githubusercontent.com/1286295/40060902-42ab50b2-581d-11e8-9981-d296e82317f5.png)

The behavior in the example above is similar to that of master, except that there are no longer any minimums for the user to set and the ratios assigned to each cache exactly match what the user specified.  In this example, 40% of the memory is devoted to bluestore metadata, 40% to RocksDB's block cache, and 20% to bluestore data.  There is not enough memory devoted to bluestore's metadata cache to cache all onode related metadata.  Once random writes start, rocksdb's block cache begins to fill with onode metadata due to cache misses at both the bluestore meta cache and rocksdb block cache levels.  In fact rocksdb is caching much of the same onode data that bluestore is caching.

![wip-bs-cache-autotune-10-85-5-false](https://user-images.githubusercontent.com/1286295/40061409-7b49cf92-581e-11e8-90c2-ddf38fd1782f.png)

Setting a larger onode ratio (85%) that is able to cache all onodes results in no double caching in rocksdb, but now we've assigned significantly more memory to bluestore's onode cache than is necessary to cache all onodes.  A significant portion of the available memory is underutilized when it could have been assigned to data cache.

# RBD Behavior -Autotuning Enabled

![wip-bs-cache-autotune-40-40-20-true](https://user-images.githubusercontent.com/1286295/40062311-8d03af30-5820-11e8-85f4-4620907965a9.png)
![wip-bs-cache-autotune-10-85-5-true](https://user-images.githubusercontent.com/1286295/40062508-ff64390a-5820-11e8-9ba9-4f0448335f26.png)

When autotuning is enabled, no matter what default ratios are specified, the autotuner converges on the same ratios.  The bluestore metadata cache is given roughly 51% of the available cache.  Because all onodes are cached in bluestore's onode cache, the KV cache is seldom utilized and remains small (~4%).  Any remaining memory is assigned to bluestore's data cache which receives a significantly larger chunk of the memory than was specified.

# RBD Performance

![wip-bs-cache-autotune-performance](https://user-images.githubusercontent.com/1286295/40069518-377b893c-5831-11e8-96e7-a68230c74aef.png)
![wip-bs-cache-autotune-99th_latency](https://user-images.githubusercontent.com/1286295/40070757-96cafb2c-5834-11e8-9bfd-6848a9eea1de.png)

The most important performance consideration for the RBD workload is avoiding bluestore meta cache misses.  The autotuner does this implicitly though it can also be accomplished by setting a high enough meta cache ratio to ensure all onodes are cached (but this is difficult for users to do!).  the Autotuner is also able to do this while setting aside a large proportion of the memory for data cache.  While this doesn't appear to dramatically help the random read results here, it may have a more  profound affect in benchmarks that stress hotspot distributions like zipf.  One notable point to make is how much keeping onodes cached improves the 99th percentile latency for reads.  Presumably as latency in the read and write path is reduced via other methods, the importance of avoiding onode cache misses is going to become more pronounced.

# RGW Behavior - Autotuning Disabled

![wip-bs-cache-autotune-40-40-20-false-rgw](https://user-images.githubusercontent.com/1286295/40067374-a42d6632-582b-11e8-9369-5726f67a0029.png)
![wip-bs-cache-autotune-10-85-5-false-rgw](https://user-images.githubusercontent.com/1286295/40067383-a888195c-582b-11e8-8419-46cc3e1a986d.png)

Due to the large amount of onode and omap data during small object RGW workloads, all caches end up filling relatively quickly.  There is still likely a large amount of onode double caching going on, competing with OMAP data in rocksdb's block cache.

# RGW Behavior - Autotuning Enabled

![wip-bs-cache-autotune-40-40-20-true-rgw](https://user-images.githubusercontent.com/1286295/40067404-afa26fe4-582b-11e8-9922-8495f3dca6fa.png)
![wip-bs-cache-autotune-10-85-5-true-rgw](https://user-images.githubusercontent.com/1286295/40067416-b32de800-582b-11e8-9edb-4b832799c739.png)

The autotuned results look much closer to the non-autotuned results for this RGW workload vs the previous RBD workload.  Because all caches end up competing for resources, the autotuner tends to converge toward the default specified ratios with one exception.  Because the autotuner prioritizes rocksdb index and filter blocks at a higher priority than typical cache items, the amount of memory devoted to rocksdb's block cache tends to increase over time.

# RGW Performance

![wip-bs-cache-autotune-performance-rgw](https://user-images.githubusercontent.com/1286295/40072207-a669025a-5838-11e8-9749-3377610f8459.png)
![wip-bs-cache-autotune-avg_latency-rgw](https://user-images.githubusercontent.com/1286295/40072212-a9dc0a04-5838-11e8-8208-a045d4ec0e0f.png)

Neither the cache ratios nor the autotuner appeared to have significant affect on the performance or latency of the RGW put workload.  A small effect may be present when memory is significantly skewed toward onode cache (to the point where indexes/filters are not cached), though it's also possible that there are other bottlenecks in the RGW path resulting in caching having minimal affect one way or the other.


# Caveats

The idea behind the cache autotuner is that it adjusts the caches relatively slowly.  By default it waits 5 seconds between rebalances and asks for memory requests in 32MB chunks (with an extra chunk of headroom if a cache is at or above capacity before trim).  This plays relatively well with bluestore's caches but rocksdb's block cache behavior is less straightforward.  Compaction may drive rocksdb to exceed it's cache ratio for a short period of time, and it's unclear how the cache balancer should handle this.  More importantly, when rocksdb exceeds the block cache size (due to compaction), all indexes and filters in the high priority pool appear to be evicted and then rapidly re-added.  It's not yet clear why this only appears to happen only when the block_cache size is exceeded and not in other scenarios.  Both of these behaviors potentially can make the autotuner thrash in undesirable ways if encountered.  To help mitigate this, a high watermark is set and memory is requested as normal at the elevated level.  If the data in the pool (high priority pool or otherwise) rapidly shrinks, the memory request size degrades slowly over time in case the cache is quickly repopulated faster than the autotuner can adapt.

# Future Work

Imagine a scenario in which a cluster receives heavy RGW writes to a large data set, then a series of small random RBD writes/reads.  Currently the autotuning mostly gives bluestore's metadata cache, data cache, and RocksDB's block cache equal priority at a single level (with the exception of RocksDB's indexes and filters).  After heavy RGW writes, each cache will likely request more memory than is available.  The autotuner will ultimately revert to the default ratios specified by the user as it has no knowledge of the relative importance of data in bluestore's onode cache and rocksdb's block cache.  A future PR may introduce the concept of binned LRU caches.  The idea being that the relative age of items in each cache should determine how much total memory each cache is assigned.  Onodes that are less than 60s old should for example be cached with a higher priority than OMAP data that is older than 60s, even if that might skew the ratios toward bluestore meta cache.  If an OSD is spending most of it's time serving RGW bucket index data, the rocksdb block cache may have significantly more recent data in it than bluestore's onode cache, in which case the ratios may skew the other direction.

Another direction for future work might be to move the cache balancer out of bluestore and make it general to the OSD itself.  Ultimately the user doesn't care what bluestore's cache size is.  They care about how much memory the OSD process uses overall.  Currently the amount of memory the OSD uses is extremely difficult to determine and is dependent on many additional factors beyond bluestore's cache size.  Future work may involve incorporating other buffers and caches into the autotuner (Pinned at PRI0) to determine how much memory each cache can be assigned.

Finally, in the RGW scenario, all 3 caches tend to fill rapidly to the point where all are competing for resources.  This results in double caching of onodes in the bluestore meta cache and rocksdb block cache.  Specifically in the block cache, recently accessed onodes and OMAP data compete for resources despite the fact that those onodes are also likely cached in bluestore's cache.  It's possible that in addition to age-binned LRU caches, we may wish to create multiple LRU caches in rocksdb for different column families of data.  IE recent OMAP data and bluestore metadata may be cached at a higher priority than recent onodes cached in rocksdb's block cache that may also be cached in the bluestore metadata cache.

Signed-off-by: Mark Nelson <mnelson@redhat.com>

markhpc commented:
retest this please

markhpc commented:
retest this please

markhpc commented:
retest this please

liewegas commented:
retest this please

markhpc commented:
retest this please

dzafman commented:
@markhpc I've seen the run-tox-ceph-disk failure on one of my build machines, but never in Jenkins.

liewegas commented:
also i think the rocksdb submodule commit needs to be moved forward.. i saw some crashes on the rocksdb ~LRUCache which implies the aligned_alloc workaround wasn't included

markhpc commented:
Ok, rebased on the newest rocksdb, fixed a corner case in the cache balancing code (we were throwing an assert when the only caches left requesting memory were those that were assigned 0% ratios) and also fixed a bug that was exposed in the LRU/2Q caches due to uint64_t subtraction being cast (and truncated) into an int.  Side benefit of all of this is that it looks like we are passing x86 and arm64 tests now, including the ceph-disk test we were previously failing for inexplicable reasons.

liewegas commented:
github is showing a conflict. rebase your branch on latest master too?

markhpc commented:
ok, rebased on master and re-added rocksdb to get rid of the submodule conflict (grr), and now make check is hitting some kind of mgr dashboard unittest failure.  I'll try retesting.

markhpc commented:
retest this please

tchaikov commented:
retest this please

tchaikov commented:
http://pulpito.ceph.com/kchai-2018-05-31_12:14:16-rados-wip-kefu-testing-2018-05-31-1806-distro-basic-smithi/

tchaikov commented:
```
2018-06-05T07:51:23.774 INFO:teuthology.orchestra.run.smithi111.stdout:[ RUN      ] ObjectStore/StoreTest.BluestoreStatistics/2
2018-06-05T07:51:25.342 INFO:teuthology.orchestra.run.smithi111.stderr:Write object
2018-06-05T07:51:25.345 INFO:teuthology.orchestra.run.smithi111.stderr:*** Caught signal (Segmentation fault) **
2018-06-05T07:51:25.345 INFO:teuthology.orchestra.run.smithi111.stderr: in thread 7fd1aa984700 thread_name:bstore_mempool
2018-06-05T07:51:25.345 INFO:teuthology.orchestra.run.smithi111.stderr: ceph version 14.0.0-260-g690f857 (690f857a8c9c5248803d5fa580020949ac7cdd17) nautilus (dev)
2018-06-05T07:51:25.346 INFO:teuthology.orchestra.run.smithi111.stderr: 1: (()+0x5a0570) [0x55b0fdef1570]
2018-06-05T07:51:25.346 INFO:teuthology.orchestra.run.smithi111.stderr: 2: (()+0x12890) [0x7fd1bd005890]
2018-06-05T07:51:25.346 INFO:teuthology.orchestra.run.smithi111.stderr: 3: (RocksDBStore::commit_cache_size()+0x94) [0x55b0fde1a5c4]
2018-06-05T07:51:25.346 INFO:teuthology.orchestra.run.smithi111.stderr: 4: (BlueStore::MempoolThread::_balance_cache(std::__cxx11::list<PriorityCache::PriCache*, std::allocator<PriorityCache::PriCache*> > const&)+0xfa) [0x55b0fdd6d52a]
2018-06-05T07:51:25.346 INFO:teuthology.orchestra.run.smithi111.stderr: 5: (BlueStore::MempoolThread::entry()+0x358) [0x55b0fdd70728]
2018-06-05T07:51:25.346 INFO:teuthology.orchestra.run.smithi111.stderr: 6: (()+0x76db) [0x7fd1bcffa6db]
2018-06-05T07:51:25.347 INFO:teuthology.orchestra.run.smithi111.stderr: 7: (clone()+0x3f) [0x7fd1b213a88f]
```

/a/kchai-2018-06-05_05:07:24-rados-wip-kefu-testing-2018-06-05-0923-distro-basic-smithi/2629444

markhpc commented:
The failure was caused due to the block_cache not being initialized when the block_cache_size is 0.  There doesn't appear to be any real reason to do this, so that code has been removed.  After doing so the test passes on my box:

```
[       OK ] ObjectStore/StoreTest.BluestoreStatistics/2 (1684 ms)
[----------] 1 test from ObjectStore/StoreTest (1684 ms total)

[----------] Global test environment tear-down
[==========] 1 test from 1 test case ran. (1684 ms total)
[  PASSED  ] 1 test.
```

I attempted to run the full ceph_test_objectstore suite but ran out of disk space in the home directory on my dev box. :)

markhpc commented:
On incerta ceph_test_objectstore hit an assert in the 2Q Cache _rm_buffer during ColSplitTest2/2.  The gist of it is that we're trying to rm a buffer and the total recorded buffer bytes is smaller than the size of the buffer to delete.  I'll take a look and track down what's going on.

```
0> 2018-06-06 10:02:52.079 7f6df6f7f980 -1 /home/perf/src/markhpc/ceph/src/os/bluestore/BlueStore.cc: In function 'virtual void BlueStore::TwoQCache::_rm_buffer(BlueStore::Buffer*)' thread 7f6df6f7f980 time 2018-06-06 10:02:52.075694
/home/perf/src/markhpc/ceph/src/os/bluestore/BlueStore.cc: 967: FAILED assert(buffer_bytes >= b->length)
```

markhpc commented:
Subsequent runs of the same test passed, so it may be difficult to reproduce.

liewegas commented:
retest this please

liewegas commented:
http://pulpito.ceph.com/sage-2018-06-11_18:21:37-rados-wip-sage3-testing-2018-06-11-0901-distro-basic-smithi/
